File: Helper/Core/Drag/ThreeFingerSwipe/ModifiedDragOutputThreeFingerSwipe.m
//
// --------------------------------------------------------------------------
// ModifiedDragOutputThreeFingerDrag.m
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2022
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#import "ModifiedDragOutputThreeFingerSwipe.h"
#import "CGSSpace.h"
#import "TouchSimulator.h"
@import Cocoa;
#import "PointerFreeze.h"
#import "Mac_Mouse_Fix_Helper-Swift.h"
#import "CGSDisplays.h"

@implementation ModifiedDragOutputThreeFingerSwipe

/// Vars

static ModifiedDragState *_drag;

static int16_t _nOfSpaces = 1;
static NSScreen *_screen = nil;

/// Interface funcs

+ (void)initializeWithDragState:(ModifiedDragState *)dragStateRef {
    _drag = dragStateRef;
}

+ (void)handleBecameInUse {

    /// Get the screen
    _screen = [NSScreen mainScreen];

    /// Get number of spaces (`_nOfSpaces`)
    ///     for use in `handleMouseInputWhileInUse()`. Getting it here for performance reasons. Not sure if significant.
    {
        /// Update [Sep 2026] Support multiple displays
        ///     Haven't measured how fast this is. Probably fast. Alternative: Iterate each space (CGSCopySpaces) and ask CGS which display it belongs to.

        NSArray *spacesInfo = CFBridgingRelease(CGSCopyManagedDisplaySpaces(CGSMainConnectionID()));
        NSString *screenUUID = [_screen mf_UUIDString];

        /// Find the entry in spacesInfo info for `_screen`
        NSDictionary *entry = nil;
        for (NSDictionary *screenDict in spacesInfo)
            if ([[screenDict objectForKey: @"Display Identifier"] isEqual: screenUUID])
                { entry = screenDict; break; }

        /// Count the spaces for `_screen`
        _nOfSpaces = [[entry objectForKey: @"Spaces"] count];
    }

    /// Freeze pointer
    if (GeneralConfig.freezePointerDuringModifiedDrag) {
        [PointerFreeze freezePointerAtPosition:_drag->usageOrigin];
    }
}

+ (void)handleMouseInputWhileInUseWithDeltaX:(double)deltaX deltaY:(double)deltaY {
    
    /**
     Horizontal dockSwipe scaling
     This makes horizontal dockSwipes (switch between spaces) follow the pointer exactly
     I arrived at these value through testing documented in the NotePlan note "MMF - Scraps - Testing DockSwipe scaling"
     TODO: Test this on a vertical screen
     */
    CGSize screenSize = _screen.frame.size;
    double originOffsetForOneSpace = _nOfSpaces <= 1 ? 2.0 : 1.0 + (1.0 / (_nOfSpaces-1));
    double spaceSeparatorWidth = 63;
    double threeFingerScaleH = originOffsetForOneSpace / (screenSize.width + spaceSeparatorWidth);
    
    /// Vertical dockSwipe scaling
    ///     Not sure if it makes sense to scale this with screen height
    double threeFingerScaleV = 1.0 / screenSize.height;
    
    /// Get phase
    
    IOHIDEventPhaseBits eventPhase = _drag->firstCallback ? kIOHIDEventPhaseBegan : kIOHIDEventPhaseChanged;
    
    /// Send events
    
    if (_drag->usageAxis == kMFAxisHorizontal) {
        double delta = -deltaX * threeFingerScaleH;
        [TouchSimulator postDockSwipeEventWithDelta:delta type:kMFDockSwipeTypeHorizontal phase:eventPhase invertedFromDevice:_drag->naturalDirection];
    } else if (_drag->usageAxis == kMFAxisVertical) {
        double delta = deltaY * threeFingerScaleV;
        [TouchSimulator postDockSwipeEventWithDelta:delta type:kMFDockSwipeTypeVertical phase:eventPhase invertedFromDevice:_drag->naturalDirection];
    }
}

+ (void)handleDeactivationWhileInUseWithCancel:(BOOL)cancel {
    
    MFDockSwipeType type;
    IOHIDEventPhaseBits phase;
    
    if (_drag->usageAxis == kMFAxisHorizontal) {
        type = kMFDockSwipeTypeHorizontal;
    } else if (_drag->usageAxis == kMFAxisVertical) {
        type = kMFDockSwipeTypeVertical;
    } else {
        assert(false);
    }
    
    phase = cancel ? kIOHIDEventPhaseCancelled : kIOHIDEventPhaseEnded;
    
    [TouchSimulator postDockSwipeEventWithDelta:0.0 type:type phase:phase invertedFromDevice:_drag->naturalDirection];
    
    /// Unfreeze pointer
    if (GeneralConfig.freezePointerDuringModifiedDrag) {
        [PointerFreeze unfreeze];
    }
    
}

+ (void)suspend {}
+ (void)unsuspend {}

@end

File: Helper/Core/Drag/ThreeFingerSwipe/ModifiedDragOutputThreeFingerSwipe.swift
//
// --------------------------------------------------------------------------
// ModifiedDragOutputThreeFingerSwipe.swift
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2022 (Translated to Swift in 2026)
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

import Foundation
import CoreGraphics
import Cocoa

@objc(ModifiedDragOutputThreeFingerSwipe)
class ModifiedDragOutputThreeFingerSwipe: NSObject, ModifiedDragOutputPlugin {
    
    private weak var drag: ModifiedDragState?
    
    private var nOfSpaces: Int16 = 1
    
    private var verticalIsUpward: Bool = false
    private var horizontalIsLeftward: Bool = false
    private var spaceSwitchSymbolicHotKeyFired: Bool = false
    
    private var appExposeSymbolicHotKeyFired: Bool = false
    private var missionControlSymbolicHotKeyFired: Bool = false
    
    func initialize(with dragState: ModifiedDragState) {
        self.drag = dragState
    }
    
    func handleBecameInUse() {
        guard let drag = self.drag else { return }
        
        // 使用宏定义的 kCGSAllSpacesMask 获取全部 Space 信息
        if let spaces = CGSCopySpaces(CGSMainConnectionID(), kCGSAllSpacesMask) {
            let uniqueSpaces = Set(spaces.takeUnretainedValue() as? [AnyHashable] ?? [])
            nOfSpaces = Int16(uniqueSpaces.count)
        }
        
        if drag.usageAxis == kMFAxisVertical {
            verticalIsUpward = drag.originOffset.y < 0
            appExposeSymbolicHotKeyFired = false
            missionControlSymbolicHotKeyFired = false
        } else if drag.usageAxis == kMFAxisHorizontal {
            horizontalIsLeftward = drag.originOffset.x < 0
            spaceSwitchSymbolicHotKeyFired = false
        }
        
        if GeneralConfig.freezePointerDuringModifiedDrag() {
            PointerFreeze.freezePointer(atPosition: drag.usageOrigin)
        }
    }
    
    func handleMouseInputWhileInUse(deltaX: Double, deltaY: Double, event: CGEvent?) {
        guard let drag = self.drag else { return }
        
        let screenSize = NSScreen.main?.frame.size ?? CGSize(width: 1920, height: 1080)
        let originOffsetForOneSpace = nOfSpaces == 1 ? 2.0 : 1.0 + (1.0 / Double(nOfSpaces - 1))
        let spaceSeparatorWidth = 63.0
        let threeFingerScaleH = originOffsetForOneSpace / (screenSize.width + spaceSeparatorWidth)
        
        let threeFingerScaleV = 1.0 / Double(screenSize.height)
        
        let beganPhase = IOHIDEventPhaseBits(kIOHIDEventPhaseBegan)
        let changedPhase = IOHIDEventPhaseBits(kIOHIDEventPhaseChanged)
        let eventPhase = drag.firstCallback ? beganPhase : changedPhase
        
        if drag.usageAxis == kMFAxisHorizontal {
            // 水平轻扫 —— 桌面空间切换
            let delta = -deltaX * threeFingerScaleH
            TouchSimulator.postDockSwipeEvent(
                withDelta: delta,
                type: kMFDockSwipeTypeHorizontal,
                phase: eventPhase,
                invertedFromDevice: drag.naturalDirection
            )
            
            if !spaceSwitchSymbolicHotKeyFired {
                spaceSwitchSymbolicHotKeyFired = true
                let shk = horizontalIsLeftward ? kMFSHMoveRightASpace : kMFSHMoveLeftASpace
                SymbolicHotKeys.post(CGSSymbolicHotKey(shk.rawValue))
            }
            
        } else if drag.usageAxis == kMFAxisVertical {
            if verticalIsUpward {
                // 向上拉起 —— Mission Control
                let delta = deltaY * threeFingerScaleV
                TouchSimulator.postDockSwipeEvent(
                    withDelta: delta,
                    type: kMFDockSwipeTypeVertical,
                    phase: eventPhase,
                    invertedFromDevice: drag.naturalDirection
                )
                
                if !missionControlSymbolicHotKeyFired {
                    missionControlSymbolicHotKeyFired = true
                    SymbolicHotKeys.post(CGSSymbolicHotKey(kMFSHMissionControl.rawValue))
                }
            } else {
                // 向下拉起 —— App Exposé
                let delta = deltaY * threeFingerScaleV
                TouchSimulator.postDockSwipeEvent(
                    withDelta: delta,
                    type: kMFDockSwipeTypeVertical,
                    phase: eventPhase,
                    invertedFromDevice: drag.naturalDirection
                )
                
                if !appExposeSymbolicHotKeyFired {
                    appExposeSymbolicHotKeyFired = true
                    SymbolicHotKeys.post(CGSSymbolicHotKey(kMFSHAppExpose.rawValue))
                }
            }
        }
    }
    
    func handleDeactivationWhileInUse(cancel: Bool) {
        guard let drag = self.drag else { return }
        
        let cancelledPhase = IOHIDEventPhaseBits(kIOHIDEventPhaseCancelled)
        let endedPhase = IOHIDEventPhaseBits(kIOHIDEventPhaseEnded)
        let phase = cancel ? cancelledPhase : endedPhase
        
        if drag.usageAxis == kMFAxisHorizontal {
            TouchSimulator.postDockSwipeEvent(
                withDelta: 0.0,
                type: kMFDockSwipeTypeHorizontal,
                phase: phase,
                invertedFromDevice: drag.naturalDirection
            )
        } else if drag.usageAxis == kMFAxisVertical {
            TouchSimulator.postDockSwipeEvent(
                withDelta: 0.0,
                type: kMFDockSwipeTypeVertical,
                phase: phase,
                invertedFromDevice: drag.naturalDirection
            )
        }
        
        if GeneralConfig.freezePointerDuringModifiedDrag() {
            PointerFreeze.unfreeze()
        }
    }
    
    func suspend() {}
    func unsuspend() {}
}
